keps/sig-api-machinery/3716-webhook-predicates/README.md

```diff
@@ -0,0 +1,865 @@
+
+With CEL match conditions, the webhook could specify wildcard match rules, and add match conditions
+to filter out the desired resources:
+
+```yaml
+rules:
+  # Match CREATE & UPDATE on all resources:
+  - operations:
+    - CREATE
+    - UPDATE
+    apiGroups: '*'
+    apiVersions: '*'
+    resources: '*'
+matchConditions:
+  # Exclude leases from the webhook
+  - expression: '!(request.resource.group == "coordination.k8s.io" && resource.resource == "leases")'
+```
+
+#### Exempt system users from security policy
+
+> As a managed cluster provider, I want to prevent user webhooks from intercepting critical system
+> requests.
+
+System _resources_ can currently be exempted through a namespace or label selector, but requests by
+system components against non-system resources cannot be. For example, update pod status requests by
+Kubelets cannot be excluded from user webhooks intercepting all pod requests.
+
+With `matchConditions`, a managed cluster could append system-exclusion rules to each webhook. For example:
+
+```yaml
+matchConditions:
+  # Exclude node requests from the webhook
+  - expression: '!("system:nodes" in request.userInfo.groups)'
+```
+
+#### Scope an NFS access management webhook to Pods mounting NFS volumes
+
+> I want to narrowly scope my webhook to only the relevant requests, in order to reduce load on the
+> webhook and reduce latency in irrelevant requests.
+
+Concrete example:
+
+> A NFS deployment uses an third-party access management system. I have an admission webhook that
+> performs an access check for against the external system for pods that mount NFS volumes. Only
+> pods with NFS volumes need to be checked.
+
+Currently, there is no way to achieve this. Many webhook implementations today start by checking
+that the request is within scope, and return early if it's not. This adds latency and an additional
+failure point to irrelevant requests. This example requires an external integration, and thus is not
+a candidate for migration to CEL `ValidatingAdmissionPolicy`.
+
+With match conditions, the expressions can check whether the request object is in-scope for the
+webhook:
+
+```yaml
+rules:
+  - operations: ['CREATE']
+    apiGroups: '' # core
+    apiVersions: '*'
+    resources: 'pods'
+matchConditions:
+  # Only include pods with an NFS volume.
+  - expression: 'request.object.spec.volumes.exists(v, v.has(nfs))'
+```
+
+### Goals
+
+1. Provide a filtering mechanism for excluding requests from an admission webhook
+2. Maintain consistency with `ValidatingAdmissionPolicy`
+
+### Non-Goals
+
+* Provide a mechanism to exclude requests from all webhooks.
+
+## Proposal
+
+### API
+
+Both `ValidatingWebhook` and `MutatingWebhook` (in `admissionregistration.k8s.io`) will be updated
+with a new `MatchConditions` field:
+
+```go
+
+type ValidatingWebhook struct {
+  // ...
+
+  // MatchConditions is a list of conditions on the AdmissionRequest ('request') that must be met for a
```

Comment:
Please add a clarifying sentence that these MatchConditions are AND'd.  Reading top to bottom I think that's the intent here.

Reply:
Done.